Docs for how to start a sandbox node and run e2e test on it (#706)
near/docs: docs/develop/contracts/sandbox.md

cd nearcore
```

Then let's build the sandbox binary, it takes several minutes depend on your CPU.

```bash
make sandbox
```

Now we can start sandbox node:

```bash
target/debug/near-sandbox --home /tmp/near-sandbox init
target/debug/near-sandbox --home /tmp/near-sandbox run
```

After you're done with sandbox node, you can stop it by `Ctrl-C`. And to clean up the data it generates, just `rm -rf /tmp/near-sandbox`.

## Run an End-to-end Test in Sandbox

Let's take status-message contract as an example. You can get this contract from: https://github.com/near/near-sdk-rs/blob/master/examples/status-message/res/status_message.wasm. This is a very straightforward contract. It has two methods:

```text
set_status(message: string)
get_status(account_id: string) -> string or null
```

`set_status` method stores message string under the sender's account name key on chain. And get_status retrieve message of given account name. If it were set by `set_status` before, it returns that string, otherwise it returns null.

1. Start a near-sandbox node. If you have run a sandbox node with some tests before, delete `/tmp/near-sandbox` before start near-sandbox.

> 2. Assume you clone the repo with status-message and source code stays in `status-message/`. The compiled contract lives in `status-message/res`. Let's do some preparation for the test:

Author: Yep good point, would require user not only just download the wasm file but clone the repo at the beginning

> 2. Assume you clone the repo with status-message and source code stays in `status-message/`. The compiled contract lives in `status-message/res`. Let's do some preparation for the test:

Author: Rewrite it to ask user clone near-sdk-rs at beginning

    new BN(10).pow(new BN(25))
  );

  const aliceUseContract = await createContractUser(
    "alice",
    config.contractAccount,
    contractMethods
  );

  const bobUseContract = await createContractUser(
    "bob",
    config.contractAccount,
    contractMethods
  );
  console.log("Finish deploy contracts and create test accounts");
  return { aliceUseContract, bobUseContract };
}

async function test() {
  // 1
  await initNear();
  const { aliceUseContract, bobUseContract } = await initTest();

  // 2
  await aliceUseContract.set_status({ args: { message: "hello" } });
  let alice_message = await aliceUseContract.get_status({
    account_id: "alice.test.near",
  });
  assert.equal(alice_message, "hello");

  // 3
  let bob_message = await bobUseContract.get_status({
    account_id: "bob.test.near",
  });
  assert.equal(bob_message, null);

  // 4
  await bobUseContract.set_status({ args: { message: "world" } });
  bob_message = await bobUseContract.get_status({
    account_id: "bob.test.near",
  });
  assert.equal(bob_message, "world");
  alice_message = await aliceUseContract.get_status({
    account_id: "alice.test.near",
  });
  assert.equal(alice_message, "hello");
}

> test();

Author: Yeah this file seems too lengthy here, put in near-sdk-rs is better, The eventual ideal place is have this implemented as some utility functions in near-api-js, to make it as short as only contains `test()` above

> test();

Author: Make the test to near/near-sdk-rs#433, once merged, will remove here

```

The test itself is very straightfoward. It does:

> 1. Create testing accounts and deploy contract.
> 2. Alice sign the transaction to set status, and get status to see set status works.
> 3. Get Bob's status and that should be null because Bob has not set status.
> 4. Bob set status and get status, should show Bob's status changed and not affect Alice status.

Reviewer: ```suggestion
1. Creates testing accounts and deploys a contract.
2. Performs a `set_status` transaction signed by Alice and then calls `get_status` to confirm `set_status` worked.
3. Gets Bob's status and which should be `null` as Bob has not yet set status.
4. Performs a `set_status` transaction signed by Bob and then calls `get_status` to show Bob's changed status and should not affect Alice's status.
```

> > Most of the code above is boilerplate setup code to setup NEAR API, key pairs, testing accounts and deploy the contract. We're working on a near-cli `near test` command to do these setup code so you can focus on writing only `test()` for this kind of test.

Reviewer: ```suggestion
> Most of the code above is boilerplate code to setup NEAR API, key pairs, testing accounts, and deploy the contract. We're working on a near-cli `near test` command to do this setup code so you can focus on writing only `test()` for this.
```

## Sandbox-only Features for Testing

> If you only use above test script that only uses standard NEAR rpcs, your tests can also be executed on a testnet node. Just replace the network ID, node url, key path and account names in above script and rerun. There's also some additional Sandbox-only features that make certain tests easier. We'll review some examples in this section.

Reviewer: ```suggestion
If you only use the above test script that just uses standard NEAR RPCs your tests can also be executed on a `testnet` node. Simply replace the network ID, node url, key path, and account names in the above script and rerun the tests. There's also some additional "Sandbox only" features that make certain tests easier. We'll review some examples of those in the following section.
```

### Patch State on the Fly

You can add or modify any contract state, contract code, account or access key during the test with `sandbox_patch_state` RPC.

> For arbitrary mutation on contract state, you cannot do it via transactions since transaction can only include contract calls that mutate state in contract programmed way. For example, for a NFT contract, you can do some operation with your owned NFTs but you cannot manipulate other people owned NFTs because smart contract has coded with checks to reject that. This is the expected behavior of the NFT contract. However you may want to change other people's NFT for test setup. This is called "arbitary mutation on contract state" and can be done by `sandbox_patch_state` RPC or stop the node, dump state as genesis, edit genesis and restart the node. The later approach is more complicated to do and also cannot be done without restart the node.

Reviewer: ```suggestion
For arbitrary mutation on contract state you cannot perform this with transactions as transactions can only include contract calls that mutate state in a contract programmed way. For example with an NFT contract, you can perform some operation with NFTs you have ownership of but you cannot manipulate NFTs that are owned by other accounts as the smart contract is coded with checks to reject that. This is the expected behavior of the NFT contract. However you may want to change another person's NFT for a test setup. This is called "arbitrary mutation on contract state" and can be done by the `sandbox_patch_state` RPC. Alternatively you can stop the node, dump state at genesis, edit genesis, and restart the node. The later approach is more complicated to do and also cannot be performed without restarting the node.
```

> For patch contract code, account or access key, you can add them with normal deploy contract, create account or add key actions in transaction, but that's also limited to your account or sub-account. `sandbox_patch_state` RPC does not have this restriction.

Reviewer: ```suggestion
For patch contract code, account, or access keys you can add them with a normal deploy contract, create account, or add key actions in a transaction but that's also limited to your account or sub-account. `sandbox_patch_state` RPC does not have this restriction.
```

> Let's see an example of how patch_state would help in a test. Assume you want to mock the real state of a mainnet, where `alice.near` has set a status, and you want to retrieve that message. Above script doesn't work out of box, because your master account is `test.near` and you can only
> create account of `alice.test.near`, not `alice.near`. Patch state can solve this problem.

Reviewer: ```suggestion
Let's explore an example of how `patch_state` would help in a test. Assume you want to mock the real state of `mainnet` where `alice.near` has set a status and you want to retrieve that status message. The above script doesn't work out of box because your master account is `test.near` and you can only create an account of `alice.test.near` not `alice.near`. Patch state can solve this problem.
```

> 1. Fetch current state from sandbox node. You can also do this with `sendJsonRpc` of `near-api-js`, or with any http client from command line:

Reviewer: ```suggestion
Here is a guide on running `patch_state`:

1. Fetch the current state from the sandbox node: _(You can also do this with `sendJsonRpc` of `near-api-js` or with any http client from command line)_
```

```bash
$ curl http://localhost:3030 -H 'content-type: application/json' -d '{"jsonrpc": "2.0", "id":1, "method":"query", "params":{"request_type":"view_state","finality":"final", "account_id":"status-message.test.near","prefix_base64":""}}'

{"jsonrpc":"2.0","result":{"values":[{"key":"U1RBVEU=","value":"AgAAAA8AAABhbGljZS50ZXN0Lm5lYXIFAAAAaGVsbG8NAAAAYm9iLnRlc3QubmVhcgUAAAB3b3JsZA==","proof":[]}],"proof":[],"block_height":24229,"block_hash":"XeCMK1jLNCu2UbkAKk1LLXEQVqvUASLoxSEz1YVBfGH"},"id":1}
```

> You can see the contract only has one key-value pair in state, looks like base64 encoded. Let's figure out what it is.

Reviewer: ```suggestion
You can see the contract only has one key-value pair in state which looks like base64 encoded. Let's figure out what it is.
```

4. Now let's add an message for `alice.near` directly to the state:

```javascript
statusMessage.records.push(new Record({ k: "alice.near", v: "hello world" }));
console.log(statusMessage);
```

5. It looks good now, let's serialize it and base64 encode it so it can be used in patch_state RPC:

```javascript
console.log(
  Buffer.from(borsh.serialize(schema, statusMessage)).toString("base64")
);
```

You'll get:

```text
AwAAAA8AAABhbGljZS50ZXN0Lm5lYXIFAAAAaGVsbG8NAAAAYm9iLnRlc3QubmVhcgUAAAB3b3JsZAoAAABhbGljZS5uZWFyCwAAAGhlbGxvIHdvcmxk
```

6. Patch state with curl:

```
curl http://localhost:3030 -H 'content-type: application/json' -d '{"jsonrpc": "2.0", "id":1, "method":"sandbox_patch_state", "params":{
  "records": [
    {
      "Data": {
        "account_id": "status-message.test.near",
        "data_key": "U1RBVEU=",
        "value": "AwAAAA8AAABhbGljZS50ZXN0Lm5lYXIFAAAAaGVsbG8NAAAAYm9iLnRlc3QubmVhcgUAAAB3b3JsZAoAAABhbGljZS5uZWFyCwAAAGhlbGxvIHdvcmxk"
      }
    }
  ]
}}'
```

> 7. Now we can back to the test file and rerun the test with new state.

Reviewer: ```suggestion
7. Now we can go back to the test file and rerun the test with the new state.
```
